src/winerror.cr

```diff
@@ -94,82 +94,125 @@ enum WinError : UInt32
-    when ERROR_WRONG_DISK                then Errno::EACCES
-    when ERROR_SHARING_BUFFER_EXCEEDED   then Errno::EACCES
-    when ERROR_BAD_NETPATH               then Errno::ENOENT
-    when ERROR_NETWORK_ACCESS_DENIED     then Errno::EACCES
-    when ERROR_BAD_NET_NAME              then Errno::ENOENT
-    when ERROR_FILE_EXISTS               then Errno::EEXIST
-    when ERROR_CANNOT_MAKE               then Errno::EACCES
-    when ERROR_FAIL_I24                  then Errno::EACCES
-    when ERROR_NO_PROC_SLOTS             then Errno::EAGAIN
-    when ERROR_DRIVE_LOCKED              then Errno::EACCES
-    when ERROR_BROKEN_PIPE               then Errno::EPIPE
-    when ERROR_DISK_FULL                 then Errno::ENOSPC
-    when ERROR_INVALID_TARGET_HANDLE     then Errno::EBADF
-    when ERROR_WAIT_NO_CHILDREN          then Errno::ECHILD
-    when ERROR_CHILD_NOT_COMPLETE        then Errno::ECHILD
-    when ERROR_DIRECT_ACCESS_HANDLE      then Errno::EBADF
-    when ERROR_SEEK_ON_DEVICE            then Errno::EACCES
-    when ERROR_DIR_NOT_EMPTY             then Errno::ENOTEMPTY
-    when ERROR_NOT_LOCKED                then Errno::EACCES
-    when ERROR_BAD_PATHNAME              then Errno::ENOENT
-    when ERROR_MAX_THRDS_REACHED         then Errno::EAGAIN
-    when ERROR_LOCK_FAILED               then Errno::EACCES
-    when ERROR_ALREADY_EXISTS            then Errno::EEXIST
-    when ERROR_INVALID_STARTING_CODESEG  then Errno::ENOEXEC
-    when ERROR_INVALID_STACKSEG          then Errno::ENOEXEC
-    when ERROR_INVALID_MODULETYPE        then Errno::ENOEXEC
-    when ERROR_INVALID_EXE_SIGNATURE     then Errno::ENOEXEC
-    when ERROR_EXE_MARKED_INVALID        then Errno::ENOEXEC
-    when ERROR_BAD_EXE_FORMAT            then Errno::ENOEXEC
-    when ERROR_ITERATED_DATA_EXCEEDS_64k then Errno::ENOEXEC
-    when ERROR_INVALID_MINALLOCSIZE      then Errno::ENOEXEC
-    when ERROR_DYNLINK_FROM_INVALID_RING then Errno::ENOEXEC
-    when ERROR_IOPL_NOT_ENABLED          then Errno::ENOEXEC
-    when ERROR_INVALID_SEGDPL            then Errno::ENOEXEC
-    when ERROR_AUTODATASEG_EXCEEDS_64k   then Errno::ENOEXEC
-    when ERROR_RING2SEG_MUST_BE_MOVABLE  then Errno::ENOEXEC
-    when ERROR_RELOC_CHAIN_XEEDS_SEGLIM  then Errno::ENOEXEC
-    when ERROR_INFLOOP_IN_RELOC_CHAIN    then Errno::ENOEXEC
-    when ERROR_FILENAME_EXCED_RANGE      then Errno::ENOENT
-    when ERROR_NESTING_NOT_ALLOWED       then Errno::EAGAIN
-    when ERROR_NO_DATA                   then Errno::EPIPE
-    when ERROR_DIRECTORY                 then Errno::ENOTDIR
-    when ERROR_NOT_ENOUGH_QUOTA          then Errno::ENOMEM
-    else                                      Errno::EINVAL
+    # https://github.com/python/cpython/blob/1446024124fb98c3051199760380685f8a2fd127/PC/errmap.h
+
+    # Unwrap FACILITY_WIN32 HRESULT errors.
+    err = self
+    err = WinError.new(err.value & 0xFFFF) if err.value & 0xFFFF0000_u32 == 0x80070000_u32
+
+    if 10000 <= err.value < 12000
```

Comment:
**question:** I understand the code structure is taken from `errmap.h` but I'm still wondering if there's a specific reason for this. Both nested `case` statemens directly compare values of `err`. They have the same shape and could be merged easily. This range comparison doesn't need to go up front. It would only be needed in the `else` branch to decide between `Errno.new(err.value.to_i32!)` and `Errno::EINVAL`.